Source/NSURL.m

```diff
@@ -2488,128 +2609,180 @@ - (NSString *) user
 }
 
 // Accessing Components in PercentEncoded Format
 - (NSString *) percentEncodedFragment
 {
-  return internal->_percentEncodedFragment;
+  return [internal->_fragment stringByAddingPercentEncodingWithAllowedCharacters:
+                    [NSCharacterSet URLFragmentAllowedCharacterSet]];
 }
 
 - (void) setPercentEncodedFragment: (NSString *)fragment
 {
-  ASSIGN(internal->_percentEncodedFragment, fragment);
+  [self setFragment: [fragment stringByRemovingPercentEncoding]];
 }
 
 - (NSString *) percentEncodedHost
 {
-  return internal->_percentEncodedHost;
+  return [internal->_host stringByAddingPercentEncodingWithAllowedCharacters:
+                    [NSCharacterSet URLHostAllowedCharacterSet]];
 }
 
 - (void) setPercentEncodedHost: (NSString *)host
 {
-  ASSIGN(internal->_percentEncodedHost, host);
+  [self setHost: [host stringByRemovingPercentEncoding]];
 }
 
 - (NSString *) percentEncodedPassword
 {
-  return internal->_percentEncodedPassword;
+  return [internal->_password stringByAddingPercentEncodingWithAllowedCharacters:
+                    [NSCharacterSet URLPasswordAllowedCharacterSet]];
 }
 
 - (void) setPercentEncodedPassword: (NSString *)password
 {
-  ASSIGN(internal->_percentEncodedPassword, password);
+  [self setPassword: [password stringByRemovingPercentEncoding]];
 }
 
 - (NSString *) percentEncodedPath
 {
-  return internal->_percentEncodedPath;
+  return [internal->_path stringByAddingPercentEncodingWithAllowedCharacters:
+                    [NSCharacterSet URLPathAllowedCharacterSet]];
 }
 
 - (void) setPercentEncodedPath: (NSString *)path
 {
-  ASSIGN(internal->_percentEncodedPath, path);
+  [self setPath: [path stringByRemovingPercentEncoding]];
 }
 
 - (NSString *) percentEncodedQuery
 {
-  return internal->_percentEncodedQuery;
+  NSString *query = @"";
+  NSEnumerator *en = [internal->_queryItems objectEnumerator];
+  NSURLQueryItem *item = nil;
+
+  while ((item = (NSURLQueryItem *)[en nextObject]) != nil)
+    {
+      NSString *name = [[item name] _stringByAddingPercentEscapesForQuery];
+      NSString *value = [[item value] _stringByAddingPercentEscapesForQuery];
+      NSString *itemString = [NSString stringWithFormat: @"%@=%@",name,value];
+      if ([query length] > 0)
+        {
+          query = [query stringByAppendingString: @"&"];
+        }
+      query = [query stringByAppendingString: itemString];
+    }
+  return query;
 }
 
 - (void) setPercentEncodedQuery: (NSString *)query
 {
-  ASSIGN(internal->_percentEncodedQuery, query);
+  [self setQuery: [query stringByRemovingPercentEncoding]];
 }
 
 - (NSArray *) percentEncodedQueryItems
 {
-  return internal->_percentEncodedQueryItems;
+  NSMutableArray *items = [NSMutableArray arrayWithCapacity: [internal->_queryItems count]];
+  NSEnumerator *en = [internal->_queryItems objectEnumerator];
+  NSURLQueryItem *i = nil;
+
+  while((i = [en nextObject]) != nil)
+    {
+      NSURLQueryItem *ni = [NSURLQueryItem queryItemWithName: [i name]
```

Comment:
I would expect that the name component needs escaping as well.

Reply:
Do you want to make these changes as I don't want to interfere with any changes you're going to make?

Reply:
I went ahead and made this simple change.  Once it's committed, I leave the branch to you.

Reply:
I am fine with either way. I just don't want to waste any more of my time on that PR just to see that another commit has made it impossible to push my changes.